File: examples/webauthn_prf.py
"""
WebAuthn PRF extension – derive symmetric secrets from a credential.

Demonstrates:
  1. Registering a credential with the PRF extension enabled
  2. Deriving a secret during authentication using eval salts
  3. Verifying that the same salts produce the same secret

The PRF extension wraps the CTAP2 hmac-secret extension.  The client
hashes application-supplied inputs through
  SHA-256("WebAuthn PRF\\0" || input)
before sending them to the authenticator as HMAC salts.

Usage: uv run python examples/webauthn_prf.py

⚠ This example creates a credential on the authenticator.
"""

from __future__ import annotations

import base64
import getpass
import json
import os
import sys

from yubikit.core.fido import FidoConnection
from yubikit.core.smartcard import SmartCardConnection
from yubikit.device import list_all_devices
from yubikit.webauthn import ClientDataCollector, UserInteraction, WebAuthnClient

ORIGIN = "https://example.com"
RP_ID = "example.com"


class ConsoleInteraction(UserInteraction):
    def prompt_up(self) -> None:
        print("\n👆 Touch your security key...")

    def request_pin(self, permissions: int, rp_id: str | None) -> str | None:
        try:
            pin = getpass.getpass("🔑 Enter PIN: ")
            return pin if pin else None
        except (EOFError, KeyboardInterrupt):
            return None

    def request_uv(self, permissions: int, rp_id: str | None) -> bool:
        return True


class SimpleCollector(ClientDataCollector):
    def collect_create(self, options_json: str) -> tuple[bytes, str]:
        options = json.loads(options_json)
        rp_id = options.get("rp", {}).get("id", RP_ID)
        client_data = json.dumps(
            {
                "type": "webauthn.create",
                "challenge": options["challenge"],
                "origin": ORIGIN,
                "crossOrigin": False,
            },
            separators=(",", ":"),
        ).encode()
        return client_data, rp_id

    def collect_get(self, options_json: str) -> tuple[bytes, str]:
        options = json.loads(options_json)
        rp_id = options.get("rpId", RP_ID)
        client_data = json.dumps(
            {
                "type": "webauthn.get",
                "challenge": options["challenge"],
                "origin": ORIGIN,
                "crossOrigin": False,
            },
            separators=(",", ":"),
        ).encode()
        return client_data, rp_id


def b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def main() -> None:
    devices = list_all_devices()
    if not devices:
        print("No YubiKeys found.", file=sys.stderr)
        sys.exit(1)

    dev, info = devices[0]
    print(f"Using: {info.serial or 'Unknown serial'}")

    conn: FidoConnection | SmartCardConnection
    if dev.supports_connection(FidoConnection):  # type: ignore[arg-type]
        conn = dev.open_connection(FidoConnection)  # type: ignore[arg-type]
    elif dev.supports_connection(SmartCardConnection):
        conn = dev.open_connection(SmartCardConnection)
    else:
        print("No usable connection.", file=sys.stderr)
        sys.exit(1)

    client = WebAuthnClient(conn, ConsoleInteraction(), SimpleCollector())

    # -- Registration with PRF enabled --
    print("\n━━━ Registration (PRF) ━━━")

    salt_input = b64url(b"example PRF salt input")

    reg_json = client.make_credential(
        json.dumps(
            {
                "rp": {"name": "PRF Example", "id": RP_ID},
                "user": {
                    "id": b64url(b"prf-user"),
                    "name": "prf@example.com",
                    "displayName": "PRF User",
                },
                "challenge": b64url(os.urandom(32)),
                "pubKeyCredParams": [{"type": "public-key", "alg": -7}],
                "timeout": 60000,
                "extensions": {"prf": {}},
            }
        )
    )

    reg = json.loads(reg_json)
    cred_id = reg["id"]
    print(f"✅ Credential: {cred_id}")

    ext = reg.get("clientExtensionResults", {})
    prf_out = ext.get("prf", {})
    if prf_out:
        print(f"  PRF enabled: {prf_out.get('enabled')}")

    # -- Authentication with PRF eval --
    print("\n━━━ Authentication (PRF eval) ━━━")

    assertions = client.get_assertion(
        json.dumps(
            {
                "challenge": b64url(os.urandom(32)),
                "timeout": 60000,
                "rpId": RP_ID,
                "allowCredentials": [{"type": "public-key", "id": cred_id}],
                "userVerification": "discouraged",
                "extensions": {
                    "prf": {
                        "eval": {
                            "first": salt_input,
                        }
                    }
                },
            }
        )
    )

    a1 = json.loads(assertions[0])
    print("✅ Authentication succeeded")
    ext1 = a1.get("clientExtensionResults", {})
    prf1 = ext1.get("prf", {}).get("results", {})
    secret1 = prf1.get("first", "")
    s1_bytes = base64.urlsafe_b64decode(secret1 + "==") if secret1 else b""
    if secret1:
        print(f"  PRF first: {s1_bytes.hex()} ({len(s1_bytes)} bytes)")

    # -- Second authentication with same salt → same secret --
    print("\n━━━ Verify determinism (same salt → same secret) ━━━")

    assertions2 = client.get_assertion(
        json.dumps(
            {
                "challenge": b64url(os.urandom(32)),
                "timeout": 60000,
                "rpId": RP_ID,
                "allowCredentials": [{"type": "public-key", "id": cred_id}],
                "userVerification": "discouraged",
                "extensions": {
                    "prf": {
                        "eval": {
                            "first": salt_input,
                        }
                    }
                },
            }
        )
    )

    a2 = json.loads(assertions2[0])
    ext2 = a2.get("clientExtensionResults", {})
    secret2 = ext2.get("prf", {}).get("results", {}).get("first", "")

    if secret1 and secret2 and secret1 == secret2:
        print(f"✅ Same salt produced the same secret ({len(s1_bytes)} bytes)")
    else:
        print("❌ Secrets differ – this is unexpected!")

    client.close()
    print("\nDone.")


if __name__ == "__main__":
    main()

File: crates/yubikit/examples/webauthn_prf.rs
// Example: WebAuthn PRF extension – derive symmetric secrets from a credential.
//
// Demonstrates:
//   1. Registering a credential with `prf` extension enabled
//   2. Deriving a secret during authentication using `eval` salts
//   3. Verifying that the same salts produce the same secret
//
// The PRF extension wraps the CTAP2 hmac-secret extension.  The client
// hashes application-supplied inputs through
//   SHA-256("WebAuthn PRF\0" || input)
// before sending them to the authenticator as HMAC salts.
//
// Usage: cargo run -p yubikit --example webauthn_prf
//
// ⚠ This example creates a credential on the authenticator.

use std::io::{self, Write};

use yubikit::ctap::CtapSession;
use yubikit::ctap2::{Ctap2Session, Permissions};
use yubikit::transport::ctaphid::{HidFidoConnection, list_fido_devices};
use yubikit::webauthn::extensions::{
    AuthenticationExtensionInputs, RegistrationExtensionInputs,
    prf::{AuthenticationInput, PrfEval, RegistrationInput},
};
use yubikit::webauthn::{
    ClientDataCollector, CollectedClientData, PublicKeyCredentialCreationOptions,
    PublicKeyCredentialDescriptor, PublicKeyCredentialParameters,
    PublicKeyCredentialRequestOptions, PublicKeyCredentialRpEntity, PublicKeyCredentialType,
    PublicKeyCredentialUserEntity, UserInteraction, UserVerificationRequirement, WebAuthnClient,
};

// ---------------------------------------------------------------------------
// Shared helpers (same as webauthn.rs)
// ---------------------------------------------------------------------------

struct ConsoleInteraction;

impl UserInteraction for ConsoleInteraction {
    fn prompt_up(&self) {
        println!("\n👆 Touch your security key...");
    }
    fn request_pin(&self, _permissions: Permissions, _rp_id: Option<&str>) -> Option<String> {
        print!("🔑 Enter PIN: ");
        io::stdout().flush().ok();
        let mut pin = String::new();
        io::stdin().read_line(&mut pin).ok()?;
        let pin = pin.trim().to_string();
        if pin.is_empty() { None } else { Some(pin) }
    }
    fn request_uv(&self, _p: Permissions, _rp: Option<&str>) -> bool {
        true
    }
}

const ORIGIN: &str = "https://example.com";
const RP_ID: &str = "example.com";

struct SimpleCollector;

impl ClientDataCollector for SimpleCollector {
    fn collect_create(
        &self,
        options: &PublicKeyCredentialCreationOptions,
    ) -> Result<(CollectedClientData, String), String> {
        let rp_id = options.rp.id.clone().unwrap_or_else(|| RP_ID.to_string());
        Ok((
            CollectedClientData::create("webauthn.create", &options.challenge, ORIGIN, false),
            rp_id,
        ))
    }
    fn collect_get(
        &self,
        options: &PublicKeyCredentialRequestOptions,
    ) -> Result<(CollectedClientData, String), String> {
        let rp_id = options.rp_id.clone().unwrap_or_else(|| RP_ID.to_string());
        Ok((
            CollectedClientData::create("webauthn.get", &options.challenge, ORIGIN, false),
            rp_id,
        ))
    }
}

fn random_challenge() -> Vec<u8> {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    use std::time::SystemTime;
    let mut h = DefaultHasher::new();
    SystemTime::now().hash(&mut h);
    std::process::id().hash(&mut h);
    let a = h.finish().to_le_bytes();
    let mut h2 = DefaultHasher::new();
    a.hash(&mut h2);
    42u64.hash(&mut h2);
    let b = h2.finish().to_le_bytes();
    [a, b, a, b].concat()
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

fn main() {
    let devices = list_fido_devices().expect("failed to list HID devices");
    let dev = devices.first().unwrap_or_else(|| {
        eprintln!("No FIDO HID devices found.");
        std::process::exit(1);
    });

    println!("Using device: {} (PID=0x{:04X})", dev.path, dev.pid);

    let conn = HidFidoConnection::open(dev).expect("failed to open HID connection");
    let ctap = match CtapSession::new_fido(conn) {
        Ok(s) => s,
        Err((e, _)) => {
            eprintln!("CTAP session error: {e}");
            std::process::exit(1);
        }
    };
    let session = Ctap2Session::new(ctap).expect("CTAP2 init failed");

    let info = session.info();
    if !info.extensions.iter().any(|e| e == "hmac-secret") {
        eprintln!("Authenticator does not support hmac-secret / PRF.");
        std::process::exit(1);
    }
    println!("✓ hmac-secret extension supported");

    let mut client = WebAuthnClient::new(session, ConsoleInteraction, SimpleCollector);

    // -- Registration with PRF enabled --
    println!("\n━━━ Registration (PRF) ━━━");
    let create_options = PublicKeyCredentialCreationOptions {
        rp: PublicKeyCredentialRpEntity {
            name: "PRF Example".to_string(),
            id: Some(RP_ID.to_string()),
        },
        user: PublicKeyCredentialUserEntity {
            id: b"prf-user".to_vec(),
            name: Some("prf@example.com".to_string()),
            display_name: Some("PRF User".to_string()),
        },
        challenge: random_challenge(),
        pub_key_cred_params: vec![PublicKeyCredentialParameters {
            type_: PublicKeyCredentialType::PublicKey,
            alg: -7,
        }],
        timeout: Some(60_000),
        exclude_credentials: None,
        authenticator_selection: None,
        hints: None,
        attestation: None,
        attestation_formats: None,
        extensions: Some(RegistrationExtensionInputs {
            prf: Some(RegistrationInput { eval: None }),
            ..Default::default()
        }),
    };

    let reg = match client.make_credential(&create_options) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Registration failed: {e}");
            std::process::exit(1);
        }
    };

    println!("✅ Credential registered: {}", hex(&reg.id));
    if let Some(ref ext) = reg.client_extension_results
        && let Some(ref prf) = ext.prf
    {
        println!("  PRF enabled: {}", prf.enabled);
    }

    // -- Authentication with PRF eval --
    println!("\n━━━ Authentication (PRF eval) ━━━");

    let salt_input = b"example PRF salt input".to_vec();

    let get_options = PublicKeyCredentialRequestOptions {
        challenge: random_challenge(),
        timeout: Some(60_000),
        rp_id: Some(RP_ID.to_string()),
        allow_credentials: Some(vec![PublicKeyCredentialDescriptor {
            type_: PublicKeyCredentialType::PublicKey,
            id: reg.id.clone(),
            transports: None,
        }]),
        user_verification: Some(UserVerificationRequirement::Discouraged),
        hints: None,
        extensions: Some(AuthenticationExtensionInputs {
            prf: Some(AuthenticationInput {
                eval: Some(PrfEval {
                    first: salt_input.clone(),
                    second: None,
                }),
                eval_by_credential: Default::default(),
            }),
            ..Default::default()
        }),
    };

    let assertions = match client.get_assertion(&get_options) {
        Ok(a) => a,
        Err(e) => {
            eprintln!("Authentication failed: {e}");
            std::process::exit(1);
        }
    };

    let first = &assertions[0];
    println!("✅ Authentication succeeded");
    if let Some(ref ext) = first.client_extension_results
        && let Some(ref prf) = ext.prf
    {
        println!("  PRF first:  {}", hex(&prf.results.first));
        if let Some(ref second) = prf.results.second {
            println!("  PRF second: {}", hex(second));
        }
    }

    // -- Second authentication with same salt → same secret --
    println!("\n━━━ Verify determinism (same salt → same secret) ━━━");

    let get_options2 = PublicKeyCredentialRequestOptions {
        challenge: random_challenge(),
        extensions: Some(AuthenticationExtensionInputs {
            prf: Some(AuthenticationInput {
                eval: Some(PrfEval {
                    first: salt_input,
                    second: None,
                }),
                eval_by_credential: Default::default(),
            }),
            ..Default::default()
        }),
        ..get_options
    };

    let assertions2 = match client.get_assertion(&get_options2) {
        Ok(a) => a,
        Err(e) => {
            eprintln!("Second auth failed: {e}");
            std::process::exit(1);
        }
    };

    let ext1 = first.client_extension_results.as_ref().unwrap();
    let ext2 = assertions2[0].client_extension_results.as_ref().unwrap();
    let s1 = &ext1.prf.as_ref().unwrap().results.first;
    let s2 = &ext2.prf.as_ref().unwrap().results.first;

    if s1 == s2 {
        println!("✅ Same salt produced the same secret ({} bytes)", s1.len());
    } else {
        println!("❌ Secrets differ – this is unexpected!");
    }

    let _session = client.into_session();
    println!("\nDone.");
}
